content/advent-2017/lexmachine-advent.md

```diff
@@ -0,0 +1,644 @@
+	lexer.Add([]byte("`"), getToken(tokmap["BACKTICK"]))
+	lexer.Add([]byte(","), getToken(tokmap["COMMA"]))
+	lexer.Add([]byte(`\(`), getToken(tokmap["LPAREN"]))
+	lexer.Add([]byte(`\)`), getToken(tokmap["RPAREN"]))
+	lexer.Add([]byte("bus"), getToken(tokmap["BUS"]))
+	lexer.Add([]byte("chip"), getToken(tokmap["CHIP"]))
+	lexer.Add([]byte("label"), getToken(tokmap["LABEL"]))
+	lexer.Add([]byte("compute"), getToken(tokmap["COMPUTE"]))
+	lexer.Add([]byte("ignore"), getToken(tokmap["IGNORE"]))
+	lexer.Add([]byte("set"), getToken(tokmap["SET"]))
+	lexer.Add([]byte(`[0-9]*\.?[0-9]+`), getToken(tokmap["NUMBER"]))
+	lexer.Add([]byte(`[a-zA-Z_][a-zA-Z0-9_]*`), getToken(tokmap["NAME"]))
+	lexer.Add([]byte(`"((\\.)|([^"\\]))*"`), func(s *lexmachine.Scanner, m *machines.Match) (interface{}, error) {
+		return s.Token(tokmap["NAME"], string(m.Bytes[1:len(m.Bytes)-1]), m), nil
+	})
+	lexer.Add([]byte(`#[^\n]*`), getToken(tokmap["COMMENT"]))
+	lexer.Add([]byte(`( |\t|\f)+`), getToken(tokmap["SPACE"]))
+	lexer.Add([]byte(`\\\n`), getToken(tokmap["SPACE"]))
+	lexer.Add([]byte(`\n|\r|\n\r`), getToken(tokmap["NEWLINE"]))
+	err = lexer.Compile()
+	if err != nil {
+		panic(err)
+	}
+	return lexer
+}
+```
+
+### Skipping a pattern
+
+Sometimes it is advantageous to not emit tokens for certain patterns and to
+instead skip them. Commonly this occurs for whitespace and comments.  To skip a
+pattern simply have the action `return nil, nil`:
+
+```go
+func skip(scan *Scanner, match *machines.Match) (interface{}, error) {
+	return nil, nil
+}
+```
+
+```go
+lexer.Add([]byte("\s+"), skip)     // skip whitespace
+lexer.Add([]byte(`#[^\n]*`), skip) // skip comments
+```
+
+In our example for `sensors.conf` we only care about the keywords and the `NAME`
+tokens. Let's skip the rest:
+
+```go
+func newLexer() *lexmachine.Lexer {
+	lexer := lexmachine.NewLexer()
+	lexer.Add([]byte(`@`), skip)
+	lexer.Add([]byte(`\+`), skip)
+	lexer.Add([]byte(`\*`), skip)
+	lexer.Add([]byte("-"), skip)
+	lexer.Add([]byte("/"), skip)
+	lexer.Add([]byte("\\"), skip)
+	lexer.Add([]byte(`\^`), skip)
+	lexer.Add([]byte("`"), skip)
+	lexer.Add([]byte(","), skip)
+	lexer.Add([]byte(`\(`), skip)
+	lexer.Add([]byte(`\)`), skip)
+	lexer.Add([]byte("bus"), skip)
+	lexer.Add([]byte("chip"), getToken(tokmap["CHIP"]))
+	lexer.Add([]byte("label"), getToken(tokmap["LABEL"]))
+	lexer.Add([]byte("compute"), skip)
+	lexer.Add([]byte("ignore"), skip)
+	lexer.Add([]byte("set"), skip)
+	lexer.Add([]byte(`[0-9]*\.?[0-9]+`), skip)
+	lexer.Add([]byte(`[a-zA-Z_][a-zA-Z0-9_]*`), getToken(tokmap["NAME"]))
+	lexer.Add([]byte(`"((\\.)|([^"\\]))*"`), func(s *lexmachine.Scanner, m *machines.Match) (interface{}, error) {
+		return s.Token(tokmap["NAME"], string(m.Bytes[1:len(m.Bytes)-1]), m), nil
+	})
+	lexer.Add([]byte(`#[^\n]*`), skip)
+	lexer.Add([]byte(`( |\t|\f)+`), skip)
+	lexer.Add([]byte(`\\\n`), skip)
+	lexer.Add([]byte(`\n|\r|\n\r`), getToken(tokmap["NEWLINE"]))
+	err = lexer.Compile()
+	if err != nil {
```

Comment:
Err not defined?

if err := lexer.Compile(); err != nil {

Reply:
yep. fixed.